app/controllers/users_controller.rb

```diff
@@ -35,4 +35,60 @@ def upload_avatar
       render 'edit'
     end
   end
+
+  # POST /user/update_email
+  def send_update_email_token_email
+    @email = current_user.email
+    @new_email = params[:new_email]
+    params[:email] = current_user.email
+    @form = Users::SendUpdateEmailForm.new(Reform::OpenForm.new)
+    @form&.current_school = current_school
+    @form.current_user = current_user
+
+    @form&.email = current_user.email
+
+    if @form.validate(params)
+      @form.save
+      render json: { success: true }
+    else
+      render json: { error: @form.errors.full_messages.join(', ') }
+    end
+  end
+
+  # GET /user/update_email
+  def update_email
+    user =
+      Users::ValidateUpdateEmailTokenService.new(params[:token], current_school)
+        .authenticate
+
+    if user.present?
+      old_email = user.email
+      new_email = user.temp_new_email
+
+      # Update user email
+      user.update!(
+        email: new_email,
+        update_email_token: nil,
+        temp_new_email: nil
+      )
+
+      # Send success email to user
+      UserMailer.confirm_email_update(user.name, user.email, current_school)
+        .deliver_now
+
+      # Send notification email to admins
```

Comment:
I think we should also log an update in [audit_record](https://github.com/pupilfirst/pupilfirst/blob/master/app/models/audit_record.rb).